acme/dns_challenge.go

```diff
@@ -62,48 +63,125 @@ func (s *dnsChallenge) Solve(chlng challenge, domain string) error {
 }
 
-func checkDNS(domain, fqdn string) bool {
-	// check if the expected DNS entry was created. If not wait for some time and try again.
+// checkDnsPropagation queries all nameservers that are authoritative for 'domain'
+// to check whether DNS propagation for the expected DNS record is complete.
+func checkDnsPropagation(domain, fqdn string) (bool, error) {
+	nameservers, err := lookupNameservers(toFqdn(domain))
+	if err != nil {
+		return false, fmt.Errorf("Could not lookup nameservers: %v", err)
+	}
+
+	for _, ns := range nameservers {
+		r, err := dnsQuery(fqdn, dns.TypeTXT, net.JoinHostPort(ns, "53"))
+		if err != nil {
+			return false, err
+		}
+
+		if r.Rcode != dns.RcodeSuccess {
+			return false, fmt.Errorf("NS %s returned RCode %s", ns, dns.RcodeToString[r.Rcode])
+		}
+
+		var found bool
+		for _, rr := range r.Answer {
+			if _, ok := rr.(*dns.TXT); ok {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false, fmt.Errorf("NS %s did not return expected TXT record", ns)
+		}
+	}
+
+	return true, nil
+}
+
+// dnsQuery directly queries the given authoritative nameserver.
+func dnsQuery(fqdn string, rtype uint16, nameserver string) (in *dns.Msg, err error) {
 	m := new(dns.Msg)
-	m.SetQuestion(domain+".", dns.TypeSOA)
-	c := new(dns.Client)
-	in, _, err := c.Exchange(m, "google-public-dns-a.google.com:53")
+	m.SetQuestion(fqdn, rtype)
+	m.SetEdns0(4096, false)
+	m.RecursionDesired = false
+
+	in, err = dns.Exchange(m, nameserver)
+	if err == dns.ErrTruncated {
+		tcp := &dns.Client{Net: "tcp"}
+		in, _, err = tcp.Exchange(m, nameserver)
+	}
+
+	return
+}
+
+// lookupNameservers returns the authoritative nameservers for the given domain name.
+func lookupNameservers(fqdn string) ([]string, error) {
+	var referralNameservers []string
+
+	// We start recursion at the gTLD origin
+	// so we don't have to manage root hints
+	labels := dns.SplitDomainName(fqdn)
+	tld := labels[len(labels)-1]
+	nss, err := net.LookupNS(tld)
 	if err != nil {
-		return false
+		return nil, fmt.Errorf("Failed to resolve TLD %s %v", tld, err)
 	}
 
-	var authorativeNS string
-	for _, answ := range in.Answer {
-		soa := answ.(*dns.SOA)
-		authorativeNS = soa.Ns
+	for _, ns := range nss {
+		referralNameservers = append(referralNameservers, ns.Host)
 	}
 
-	fallbackCnt := 0
-	for fallbackCnt < preCheckDNSFallbackCount {
-		m.SetQuestion(fqdn, dns.TypeTXT)
-		in, _, err = c.Exchange(m, authorativeNS+":53")
-		if err != nil {
-			return false
+	// Follow the referrals until we hit the authoritative NS
+	for depth := 0; depth < recursionMaxDepth; depth++ {
+		var r *dns.Msg
+		var err error
+		
+		for _, ns := range referralNameservers {
+			r, err = dnsQuery(fqdn, dns.TypeNS, net.JoinHostPort(ns, "53"))
+			if err != nil {
+				continue
+			}
+			if r.Rcode == dns.RcodeSuccess {
+				break
+			}
+			if r.Rcode == dns.RcodeNameError {
```

Comment:
In my use case, `internal-name.willglynn.com` isn't nonexistent, it's just not visible to public resolvers due to [split horizon DNS](https://en.wikipedia.org/wiki/Split-horizon_DNS). The `_acme-challenge` record obviously needs to be visible, but there's no constraint on `internal-name` itself, hence `lookupNameservers()` queries returning `NXDOMAIN`.

As for spinning forever, I see the timeout-related code, but… `lego` does *not* time out after a minute in this branch. I'm not sure where it's getting stuck (nor is there any obvious way to get verbose output), but after several minutes of waiting, `tshark` shows my machine repeatedly querying my local recursive resolver for `com. IN NS`:

    2016/02/07 01:25:32 [INFO][internal-name.willglynn.com] acme: Obtaining bundled SAN certificate
    2016/02/07 01:25:32 [INFO][internal-name.willglynn.com] acme: Could not find solver for: http-01
    2016/02/07 01:25:32 [INFO][internal-name.willglynn.com] acme: Could not find solver for: tls-sni-01
    2016/02/07 01:25:32 [INFO][internal-name.willglynn.com] acme: Trying to solve DNS-01
    … wait a long time …
    … switch terminals …
    $ tshark -t ad -i en0 'port 53'
    Capturing on 'Wi-Fi'
      1 2016-02-07 01:34:42.914351 172.16.3.100 -> 172.16.3.1   DNS 63 Standard query 0xc0f5 NS com
      2 2016-02-07 01:34:42.920370   172.16.3.1 -> 172.16.3.100 DNS 565 Standard query response 0xc0f5 NS m.gtld-servers.netNS c.gtld-servers.netNS d.gtld-servers.netNS b.gtld-servers.netNS e.gtld-servers.netNS h.gtld-servers.netNS g.gtld-servers.netNS l.gtld-servers.netNS f.gtld-servers.netNS a.gtld-servers.netNS k.gtld-servers.netNS i.gtld-servers.netNS j.gtld-servers.net
      3 2016-02-07 01:34:42.920926 172.16.3.100 -> 172.16.3.1   DNS 63 Standard query 0x4ef1 NS com
      4 2016-02-07 01:34:42.927413   172.16.3.1 -> 172.16.3.100 DNS 565 Standard query response 0x4ef1 NS d.gtld-servers.netNS b.gtld-servers.netNS e.gtld-servers.netNS h.gtld-servers.netNS g.gtld-servers.netNS l.gtld-servers.netNS f.gtld-servers.netNS a.gtld-servers.netNS k.gtld-servers.netNS i.gtld-servers.netNS j.gtld-servers.netNS m.gtld-servers.netNS c.gtld-servers.net